elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +306,319 @@ private Object fetchRelationship(Environment context) {
+        private Triplet(Map<String, Object> input, Optional<PersistentResource> parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * A function to handle upserting (update/create) objects.
+     */
+    @FunctionalInterface
+    private interface Executor<T> {
+        T execute(Map<String, Object> input, String id,
+                                   RequestScope requestScope, Class<?> entityClass,
+                                   String fieldName, Optional<PersistentResource> parent);
+    }
+
+    /**
+     * Forms the graph from data {@param input} and upsert's {@param function} all the nodes
+     * @param input input data
+     * @param function function to process nodes
+     * @param parent parent resource
+     * @param requestScope request scope
+     * @param fieldName graphql field name
+     * @return set of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> graphWalker(Map<String, Object> input, Executor function,
+                                                Optional<PersistentResource> parent, RequestScope requestScope,
+                                                String fieldName) {
+        EntityDictionary dictionary = requestScope.getDictionary();
+        Queue<Triplet> toVisit = new ArrayDeque();
+        Set<Triplet> visited = new HashSet();
+        toVisit.add(new Triplet(input, parent, fieldName));
+        Set<PersistentResource> toReturn = new HashSet<>();
+
+        while(!toVisit.isEmpty()) {
+            Triplet triplet = toVisit.remove();
+            if(visited.contains(triplet)) {
+                continue;
+            } else {
+                visited.add(triplet);
+            }
+            Class<?> entityClass = getEntityClass(triplet.parentResource, dictionary, triplet.fieldName);
+            String idFieldName = getIdFieldName(dictionary, entityClass);
+            Optional<String> id = getId(triplet.input, idFieldName);
+            if(!id.isPresent()) {
+                setId(triplet.input, idFieldName);
+            }
+            /* first execute the given function with input data and then add the remaining relationships back to queue */
+            PersistentResource newParent = (PersistentResource) function.execute(triplet.input, getId(triplet.input,
+                    idFieldName).get(), requestScope, entityClass, triplet.fieldName, triplet.parentResource);
+            toReturn.add(newParent);
+            Map<String, Object> relationshipsOnly = stripAttributes(triplet.input, entityClass,
+                    dictionary);
+            /* loop over relationships */
+            for(Map.Entry<String, Object> relationship : relationshipsOnly.entrySet()) {
+                Boolean isToOne = newParent.getRelationshipType(relationship.getKey()).isToOne();
+                List<Map<String, Object>> entryToAdd;
```

Comment:
This section could be rewritten simpler like:

List<Map<String, Object>> entryToAdd = new ArrayList<>();
if (isToOne) {
    entryToAdd.add((Map<String, Object>) relationship.getValue());
} else {
    entryToAdd.addAll((List) relationship.getValue());
}
```